src/library/scala/collection/LinearSeq.scala

```diff
@@ -220,6 +220,19 @@ trait LinearSeqOps[+A, +CC[X] <: LinearSeq[X], +C <: LinearSeq[A] with LinearSeq
     last
   }
 
+  override def findLast(p: A => Boolean): Option[A] = {
+    val marker = new AnyRef
+    var these: LinearSeq[A] = coll
+    var last: AnyRef = marker // don't use `Option`, to prevent excessive `Some` allocation
+    while (these.nonEmpty) {
+      val elem = these.head
+      if (p(elem)) last = elem.asInstanceOf[AnyRef]
```

Comment:
This changes the implicit contract of `findLast` from evaluating `p` only on the tail to evaluating `p` on everything.  I'm not sure this is a good idea given that some people may rely upon side-effects, and there's no way to tell that a subclass might randomly decide to evaluate things in a different order.  At the very least, for common results, changing performance from O(1) to O(n) on an expensive p could be really surprising.

(Also, I bet using a Boolean flag instead of an allocated marker is faster 99% of the time.  If you made the marker package-private and stuck it in some object somewhere that might be faster yet, but in my tests boolean flags are usually really fast.)

Reply:
I agree that it's not necessarily a good idea to evaluate the `p` on all elements; however that is also the behaviour of `lastIndexWhere`. They should probably both be changed.